In [4]:
import numpy as np
def f(x):
    return x**2 - 4*x + 5

def grad(x):
    return 2*x - 4

eta = 0.2
x = 5

print("Bước 0: x =", x, "f(x) =", f(x))
for i in range(4):
    g = grad(x)
    x = x - eta * g
    print("Bước", i + 1, ": x =", x, "f(x) =", f(x))

Bước 0: x = 5 f(x) = 10
Bước 1 : x = 3.8 f(x) = 4.24
Bước 2 : x = 3.08 f(x) = 2.1663999999999994
Bước 3 : x = 2.648 f(x) = 1.4199039999999998
Bước 4 : x = 2.3888000000000003 f(x) = 1.1511654399999998


# Nhận xét:
- x giảm dần từ 5 xuống 2.3888.
- f(x) giảm từ 10 xuống 1.1511, x đang tiến gần đến 2
- thuật toán Gradient Descent đang hội tụ.